contracts/AddressBook.sol

```diff
@@ -3,10 +3,116 @@
+    }
+
+    function getOtokenFactory() external view returns (address) {
+        return getAddress(OTOKEN_FACTORY);
+    }
+
+    function getWhitelist() external view returns (address) {
+        return getAddress(MARGIN_WHITELIST);
+    }
+
+    function getController() external view returns (address) {
+        return getAddress(CONTROLLER);
+    }
+
+    function getMarginPool() external view returns (address) {
+        return getAddress(MARGIN_POOL);
+    }
+
+    function getMarginCalculator() external view returns (address) {
+        return getAddress(MARGIN_CALCULATOR);
+    }
+
+    function getLiquidationManager() external view returns (address) {
+        return getAddress(LIQUIDATION_MANAGER);
+    }
+
+    function getOracle() external view returns (address) {
+        return getAddress(ORACLE);
+    }
+
+    function setAddress(bytes32 _key, address _module) public onlyOwner {
+        modules[_key] = _module;
+    }
+
+    function setOtokenImpl(address _otokenImpl) external onlyOwner {
+        updateImplInternal(OTOKEN_IMPL, _otokenImpl);
+    }
+
+    function setOtokenFactory(address _otokenFactory) external onlyOwner {
+        updateImplInternal(OTOKEN_FACTORY, _otokenFactory);
+    }
+
+    function setWhitelist(address _whitelist) external onlyOwner {
+        updateImplInternal(MARGIN_WHITELIST, _whitelist);
+    }
+
+    function setController(address _controller) external onlyOwner {
+        updateImplInternal(CONTROLLER, _controller);
+    }
+
+    function setMarginPool(address _marginPool) external onlyOwner {
+        updateImplInternal(MARGIN_POOL, _marginPool);
+    }
+
+    function setMarginCalculator(address _marginCalculator) external onlyOwner {
+        updateImplInternal(MARGIN_CALCULATOR, _marginCalculator);
+    }
+
+    function setLiquidationManager(address _liquidationManager) external onlyOwner {
+        updateImplInternal(LIQUIDATION_MANAGER, _liquidationManager);
+    }
+
+    function setOracle(address _oracle) external onlyOwner {
+        updateImplInternal(ORACLE, _oracle);
+    }
+
+    /**
+     * @dev internal function to update the implementation of a specific component of the protocol
+     * @param _id the id of the contract to be updated
+     * @param _newAddress the address of the new implementation
+     **/
+    function updateImplInternal(bytes32 _id, address _newAddress) public onlyOwner {
+        address payable proxyAddress = address(uint160(getAddress(_id)));
+
+        OwnedUpgradeabilityProxy proxy = OwnedUpgradeabilityProxy(proxyAddress);
+        bytes memory params = abi.encodeWithSignature("initialize(address)", address(this));
 
+        if (proxyAddress == address(0)) {
+            proxy = new OwnedUpgradeabilityProxy();
+            proxy.upgradeToAndCall(_newAddress, params);
+            setAddress(_id, address(proxy));
+            emit ProxyCreated(_id, address(proxy));
+        } else {
+            proxy.upgradeToAndCall(_newAddress, params);
```

Comment:
So this is setting a particular implementation and will only be called if the implementation is upgraded / a new contract not in the module is created?

Reply:
yeah basically this function will get a module key and a deployed address, and will create an upgradeable proxy for it, if there is no V1 deployed, otherwise will upgrade V1 implementation to the new one (V2).

The `describe('Upgrade contract')` in the tests explain more the upgrading flow, and will share a doc soon about it.